File: algorithm/16. 3Sum Closest/16.cpp
class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        int result=0,i=0,min_diff=INT_MAX;
        sort(nums.begin(),nums.end());
        while (i < nums.size()-2)
        {
            if(i==0 || nums[i]!=nums[i-1])
            {
                int j=i+1, k = nums.size()-1;
                while(j<k)
                {
                    int diff=nums[i]+nums[j]+nums[k]-target;
                    if(abs(diff)<min_diff)
                    {
                        min_diff=abs(diff);
                        result= nums[i]+nums[j]+nums[k];
                    }
                    if(diff>0)
                    {
                        k--;
                    }
                    else if(diff<0)
                        j++;
                    else
                        return target;
                }
            }
            i++;
        }
        return result;
    }
};

File: algorithm/16. 3Sum Closest/16.java
class Solution {
    public int threeSumClosest(int[] nums, int target) {
        int min_diff = Integer.MAX_VALUE, result = 0;
        Arrays.sort(nums);
        for(int i = 0; i < nums.length; i++)
        {
            if(i > 0 && nums[i] == nums[i - 1]) continue;
            int left = i + 1, right = nums.length - 1;
            while(left < right)
            {
                int diff = nums[i] + nums[left] + nums[right] - target;
                if(Math.abs(diff) < min_diff)
                {
                    min_diff = Math.abs(diff);
                    result = nums[i] + nums[left] + nums[right];
                }
                if(diff < 0)
                    left++;
                else if(diff > 0)
                    right--;
                else
                    return target;
            }
        }
        return result;
    }
}
